pandas/core/algorithms.py

```diff
@@ -753,38 +758,58 @@ def factorize(
 
     elif not isinstance(values.dtype, np.dtype):
         if (
-            na_sentinel == -1
-            and "use_na_sentinel" in inspect.signature(values.factorize).parameters
-        ):
+            na_sentinel == -1 or na_sentinel is None
+        ) and "use_na_sentinel" in inspect.signature(values.factorize).parameters:
             # Avoid using catch_warnings when possible
             # GH#46910 - TimelikeOps has deprecated signature
             codes, uniques = values.factorize(  # type: ignore[call-arg]
-                use_na_sentinel=True
+                use_na_sentinel=na_sentinel is not None
```

Comment:
Yes - but how much depends on what you mean by "this". There is a lot of playing around with arguments (what I think you're highlighting here) that will all go away. But what I regard as the main complication this PR introduces, noted in the TODO immediately below, will not go away just with the deprecation.

For that TODO, some changes to `safe_sort` are necessary. For the bulk of cases the changes are straightforward and only require very small changes. However if you have multiple Python types in an object array (e.g. float and string) with np.nan and None, I haven't yet to find a good way to sort. I plan to revisit this in the next few days.

Reply:
I looked at safe_sort again, and @phofl already solve much of the issue in #47331. I opened a PR into this branch here to see what the execution of the TODO mentioned above would look like: https://github.com/rhshadrach/pandas/pull/2

However, changing safe_sort in this way will induce another behavior change:

```
df = pd.DataFrame({'a': ['x', None, np.nan], 'b': [1, 2, 3]})
print(df.groupby('a', dropna=False).sum())

# main
     b
a     
x    1
NaN  5

# feature
      b
a      
x     1
None  2
NaN   3
```

No other tests besides the one changed failed locally. While I do think this is a bugfix, I'd like to study more what impact it has on concat/reshaping/indexing and I think it may need some discussion. In particular, I don't think it should be done in this PR.

Reply:
@jbrockmendel - friendly ping.

Reply:
> While I do think this is a bugfix, I'd like to study more what impact it has on concat/reshaping/indexing and I think it may need some discussion. In particular, I don't think it should be done in this PR.

Agreed both that the "feature" behavior looks more correct and that it should be done separate from this PR.

Taking another look now.

Reply:
(Coming in fairly cold), I think I'm getting lost too here. I am not fully comprehending why we check `            na_sentinel == -1 or na_sentinel is None` and then use `use_na_sentinel=na_sentinel is not None`

Reply:
Yea - there are some gymnastics here no doubt. The idea behind this block is to avoid using catch_warnings since it's possible.

Old API: na_sentinel is either an integer or None
New API: use_na_sentinel is False or True

The correspondence is:

use_na_sentinel False is equivalent to na_sentinel being None
use_na_sentinel True is equivalent to na_sentinel is -1

Note there is no option in the new API for na_sentinel being anything other than -1 or None in the old API. So we can use the new argument precisely when (a) the function has said argument and (b) na_sentinel is either -1 or None. In such a case, the correspondence from na_sentinel to use_na_sentinel is given by `use_na_sentinel = na_sentinel is not None`.

Reply:
Perfect, just the explanation I needed. Thanks!

Reply:
I'll add this correspondence as a comment to the top of this function; we can remove it when the deprecation is enforced.